src/syntaxtree.c

```diff
@@ -240,15 +331,70 @@ static Obj SyntaxTreeRecExpr(Obj result, Expr expr)
     return result;
 }
 
-static Obj SyntaxTreeFloatLazy(Obj result, Expr expr)
+static Expr SyntaxTreeCodeRecExpr(Obj node)
 {
-    Obj  string;
+    Obj   keyvalue = ELM_REC(node, RNamName("keyvalue"));
+    UInt  len = LEN_LIST(keyvalue);
+    UInt1 tnum = FIND_TNUM_FROM_STRING(ELM_REC(node, RNamName("type")));
+    Expr  record = NewStatWithProf(tnum, 2 * len * sizeof(Expr), 0);
+    for (int i = 0; i < len; i++) {
+        Obj  keyvaluepair = ELM_PLIST(keyvalue, i + 1);
+        Obj  keynode = ELM_REC(keyvaluepair, RNamName("key"));
+        Obj  valuenode = ELM_REC(keyvaluepair, RNamName("value"));
+        Expr key;
+        if (IS_STRING(keynode)) {
+            key = INTEXPR_INT(RNamObj(keynode));
+        }
+        else {
+            key = SyntaxTreeDefaultCoder(keynode);
+        }
+        Expr value = SyntaxTreeDefaultCoder(valuenode);
+        WRITE_EXPR(record, 2 * i, key);
+        WRITE_EXPR(record, 2 * i + 1, value);
+    }
+    return record;
+}
 
-    string = GET_VALUE_FROM_CURRENT_BODY(READ_EXPR(expr, 1));
+static Obj SyntaxTreeFloatLazy(Obj result, Expr expr)
+{
+    Obj string = GET_VALUE_FROM_CURRENT_BODY(READ_EXPR(expr, 1));
     AssPRec(result, RNamName("value"), string);
     return result;
 }
 
+static Obj SyntaxTreeFloatEager(Obj result, Expr expr)
+{
+    Obj value = GET_VALUE_FROM_CURRENT_BODY(READ_EXPR(expr, 0));
+    Obj string = GET_VALUE_FROM_CURRENT_BODY(READ_EXPR(expr, 1));
+    AssPRec(result, RNamName("value"), value);
+    AssPRec(result, RNamName("string"), string);
+    return result;
+}
+
+static Expr SyntaxTreeCodeFloatLazy(Obj node)
+{
+    Obj  value = ELM_REC(node, RNamName("value"));
+    Expr fl = NewStatWithProf(T_FLOAT_EXPR_LAZY, 2 * sizeof(UInt), 0);
+    UInt ix = CheckForCommonFloat(CONST_CSTR_STRING(value));
+    if (!ix) {
+        ix = getNextFloatExprNumber();
+    }
+    WRITE_EXPR(fl, 0, ix);
+    WRITE_EXPR(fl, 1, PushValue(value));
+    return fl;
```

Comment:
This is 90% identical to `CodeLazyFloatExpr`. I wonder if we should just call that, and then use `PopExpr`. Or alternatively, extract the common code into a helper. I'd prefer that over exporting the super low-level `getNextFloatExprNumber`, which is a (problematic) implementation detail that I hope will go away at some point...

Reply:
I added the additional parameter.